### PR TITLE
Tkv limit bug

### PR DESCRIPTION
**Pull request by Daniel-Schenker**
## Description

This PR addresses a bug in the scheduler code where the scheduler would allow requests that would eventually exceed the max batch tkv limit during an ongoing batch. The PR includes a new test case to reproduce the error and the fix for said error.

## Test Plan

Run the test file with: `pytest tests/v1/worker/test_scheduler_tkv_limits.py -sv`
Second test case will pass with changes in `sendnn_inference/v1/core/scheduler.py` included in this PR, and will fail without.

## Checklist

- [x] I have read the [contributing guidelines](https://docs.vllm.ai/projects/spyre/en/latest/contributing)
- [x] My code follows the project's code style (run `bash format.sh`)
- [x] I have added tests for my changes (if applicable)
- [ ] I have updated the documentation (if applicable)
- [x] My commits include a `Signed-off-by:` line (DCO compliance)

### CODE REVIEW

**Comment by Daniel-Schenker** [comment]
Initial proposed changes are here. More than happy to address any comments and add/change code and comments if things aren't clear.

File: sendnn_inference/v1/core/scheduler.py [hunk 1]
```diff
@@ -21,6 +21,13 @@
 logger = init_logger(__name__)
 
 
+def round_up_to_block_size(n: int) -> int:
+    # Helper function to round up to the nearest block size
+    # Uses bitwise alignment for better performance
+    block_size = SpyrePlatform.get_block_size()
+    return (n + block_size - 1) & ~(block_size - 1)
```
Comment (joerunde) [inline sendnn_inference/v1/core/scheduler.py]: I'm not sure we should make this generic, because the bitwise ops only work if the block size is a power of 2. What about hardcoding 64 here, and sticking an assert in the calling code that block_size == 64?

I have a reasonably low care amount because we never expect the block size to change, only a slight preference to not leave a foot-gun lying around.

File: tests/v1/worker/test_scheduler_tkv_limits.py [hunk 2]
```diff
@@ -80,3 +80,135 @@ def test_scheduler_tkv_limits(monkeypatch: pytest.MonkeyPatch):
...
+    )
+
+    # Configure the TKV limit
+    scheduler = model_runner.scheduler
+    scheduler.max_batch_tkv_limit = 131072
+    SpyrePlatform._max_batch_tkv_limit = 131072
+    monkeypatch.setenv("VLLM_DT_MAX_BATCH_TKV_LIMIT", "131072")
+
+    # Define prompt lengths for first set of requests
+    prompt_lengths_1 = [
```
Comment (joerunde) [inline tests/v1/worker/test_scheduler_tkv_limits.py]: [1024] * 16?

or [1018] + [1024] * 15 if the first 1018 is relevant

File: sendnn_inference/v1/core/scheduler.py [hunk 3]
```diff
@@ -531,17 +538,18 @@ def check_batch_tkv_limit_cp(
         """
 
         # Compute the effective token length of the new request
-        new_req_max_tkv = new_req_tkv + request.max_tokens - 1
+        # Rounded up to the nearest block size
```
Comment (joerunde) [inline sendnn_inference/v1/core/scheduler.py]: Should we mention here (and below) that we're rounding up to the end of the block to account for padding?

File: tests/v1/worker/test_scheduler_tkv_limits.py [hunk 4]
```diff
@@ -80,3 +80,135 @@ def test_scheduler_tkv_limits(monkeypatch: pytest.MonkeyPatch):
         scheduler.update_from_output(sched_output, output)
         if len(scheduler.running) == 0:
             break
+
+
+@pytest.mark.cpu
+@pytest.mark.chunked_prefill
+def test_scheduler_tkv_limits_ongoing_batch(monkeypatch: pytest.MonkeyPatch):
```
Comment (joerunde) [inline tests/v1/worker/test_scheduler_tkv_limits.py:87]: A high-level overview of what's going on here would be great.

IIUC this situation is:
- We first schedule a 16x8k batch that would fully fill the 128k TKV limit
- We then inject a batch of smaller requests partway through processing, which should be able to schedule only because they are guaranteed to finish processing just before the TKV is long enough to overrun the limit with the larger batch size
- This flexes the logic for injecting shorter requests into a running batch, which is not tested by the other test in this file

File: sendnn_inference/v1/core/scheduler.py (hunk 1 above)
Comment (Daniel-Schenker) [reply sendnn_inference/v1/core/scheduler.py]: Addressed

File: tests/v1/worker/test_scheduler_tkv_limits.py (hunk 2 above)
Comment (Daniel-Schenker) [reply tests/v1/worker/test_scheduler_tkv_limits.py]: Addressed, also realized that the second set of prompt lengths was identical to the first. Only difference between the 2 sets of requests is the max tokens so consolidated that as well.

File: sendnn_inference/v1/core/scheduler.py (hunk 3 above)
Comment (Daniel-Schenker) [reply sendnn_inference/v1/core/scheduler.py]: Addressed

File: tests/v1/worker/test_scheduler_tkv_limits.py (hunk 4 above)
Comment (Daniel-Schenker) [reply tests/v1/worker/test_scheduler_tkv_limits.py:87]: Added

**Comment by Daniel-Schenker** [comment]
Initial comments have been addressed. I have re-confirmed that the new test case fails without scheduler.py changes and passes with changes.

File: sendnn_inference/v1/core/scheduler.py [hunk 5]
```diff
@@ -20,6 +20,16 @@
...
 
+# Ensure that block_size is 64
+# This ensures the rounding function is correct
+assert SpyrePlatform.get_block_size() == 64
+
+
+def round_up_to_block_size(n: int) -> int:
+    # Helper function to round up to the nearest block size
+    # Uses bitwise alignment for better performance
+    return (n + 63) & ~63
```
Comment (yannicks1) [inline sendnn_inference/v1/core/scheduler.py:31]: spicy stuff 🌶️

File: sendnn_inference/v1/core/scheduler.py
Comment (yannicks1) [inline sendnn_inference/v1/core/scheduler.py]: `n_blocks` is unused, please remove

**Review by yannicks1 — COMMENTED** [review]
To me it looks like this bug was (at least partially) introduced in https://github.com/torch-spyre/sendnn-inference/pull/913 (please correct me if I am wrong). Prior to 913 we passed `n_blocks` into the function which had information about the new request and did calculation based on this. While we fixed an edge case in 913, the new approach did not use `n_blocks`, hence the new request was not considered. 

With this PR we consider the new request again while still fixing the edge case bug.

**Comment by yannicks1** [comment]
> Initial comments have been addressed. I have re-confirmed that the new test case fails without scheduler.py changes and passes with changes.

could you quickly run your newly added test (that captured the new failure) with a pre 913 commit? This would help me understand better...

File: sendnn_inference/v1/core/scheduler.py
Comment (Daniel-Schenker) [reply sendnn_inference/v1/core/scheduler.py]: Removed

Comment (yannicks1) [inline sendnn_inference/v1/core/scheduler.py]: this needs to be gone too ..

Comment (Daniel-Schenker) [reply sendnn_inference/v1/core/scheduler.py]: Removed

**Review by joerunde — APPROVED** [review]
ship it!
